polygeist/lib/Dialect/Polygeist/Analysis/SYCLAccessorAnalysis.cpp

```diff
@@ -0,0 +1,403 @@
+  SmallVector<size_t, 3> jointOffset = (constantOffset == other.constantOffset)
+                                           ? constantOffset
+                                           : SmallVector<size_t, 3>{};
+
+  return AccessorInformation(jointBuf, jointBufInfo, jointNeedsRange,
+                             jointRange, jointNeedsOffset, jointOffset);
+}
+
+AliasResult AccessorInformation::alias(const AccessorInformation &other,
+                                       AliasAnalysis &aliasAnalysis) const {
+  // If we can't analyze the underlying buffer, can't determine aliasing, so
+  // must assume they may alias.
+  if (!hasKnownBuffer() || !other.hasKnownBuffer())
+    return AliasResult::MayAlias;
+
+  auto isSameValue = [&](Value one, Value two) -> bool {
+    return one == two ||
+           aliasAnalysis.alias(one, two) == AliasResult::MustAlias;
+  };
+
+  if (isSameValue(getBuffer(), other.getBuffer())) {
+    // Try to refine to must alias
+    if (!needsRange() && !other.needsRange())
+      // If neither uses a range (and therefore also no offset), they must
+      // alias.
+      return AliasResult::MustAlias;
+
+    // Both accessors defined on the same buffer.
+    if (!hasBufferInformation() || (needsRange() && !hasConstantRange()) ||
+        (needsOffset() && !hasConstantOffset()) ||
+        !other.hasBufferInformation() ||
+        (other.needsRange() && !other.hasConstantRange()) ||
+        (other.needsOffset() && !other.hasConstantOffset()))
+      // Without definitive information about the underlying buffer and the
+      // range & offset of the two accessors, they might be set up in way
+      // where they completely or partially overlap, so may alias.
+      return AliasResult::MayAlias;
+
+    if (!needsRange()) {
+      // This accessor covers the entire range of the buffer.
+      if (other.hasConstantRange() && getBufferInfo().hasConstantSize() &&
+          SmallVector<size_t, 3>(other.getConstantRange()) ==
+              SmallVector<size_t, 3>(getBufferInfo().getConstantSize()) &&
+          !other.needsOffset())
+        // The other accessor also covers the entire buffer, so they must alias
+        return AliasResult::MustAlias;
+
+      // The other accessor only covers part of the buffer, so they alias for
+      // some elements.
+      return AliasResult::MayAlias;
+    }
+
+    if (!other.needsRange()) {
+      // The other accessor covers the entire range of the buffer.
+      if (hasConstantRange() && getBufferInfo().hasConstantSize() &&
+          SmallVector<size_t, 3>(getConstantRange()) ==
+              SmallVector<size_t, 3>(getBufferInfo().getConstantSize()) &&
+          !needsOffset())
+        // This accessor also covers the entire buffer, so they must alias
+        return AliasResult::MustAlias;
+
+      // This accessor only covers part of the buffer, so they alias for
+      // some elements.
+      return AliasResult::MayAlias;
```

Comment:
This code is similar to lines 134-146. Possible to create a utility function or lambda to common up the code ?